apps/api/src/app/events/usecases/trigger-event/trigger-event.usecase.ts

```diff
@@ -1,65 +1,109 @@
 import { StepTypeEnum, ExecutionDetailsSourceEnum, ExecutionDetailsStatusEnum } from '@novu/shared';
 import * as Sentry from '@sentry/node';
 
 import { TriggerEventCommand } from './trigger-event.command';
-import { ProcessSubscriber } from '../process-subscriber/process-subscriber.usecase';
-import { ProcessSubscriberCommand } from '../process-subscriber/process-subscriber.command';
-import { ApiException } from '../../../shared/exceptions/api.exception';
+
 import { AddJob } from '../add-job/add-job.usecase';
+import { CreateNotificationCommand, CreateNotification } from '../create-notification';
+import { ProcessSubscriber, ProcessSubscriberCommand } from '../process-subscriber';
+
 import { CreateExecutionDetails } from '../../../execution-details/usecases/create-execution-details/create-execution-details.usecase';
 import {
   CreateExecutionDetailsCommand,
   DetailEnum,
 } from '../../../execution-details/usecases/create-execution-details/create-execution-details.command';
+import { ApiException } from '../../../shared/exceptions/api.exception';
+import { Cached } from '../../../shared/interceptors';
+import { CacheKeyPrefixEnum } from '../../../shared/services/cache';
+
+const LOG_CONTEXT = 'TriggerEventUseCase';
 
 @Injectable()
 export class TriggerEvent {
   constructor(
-    private notificationTemplateRepository: NotificationTemplateRepository,
+    private addJobUsecase: AddJob,
+    private createNotification: CreateNotification,
     private processSubscriber: ProcessSubscriber,
     private jobRepository: JobRepository,
-    private addJobUsecase: AddJob,
     private notificationRepository: NotificationRepository,
+    private notificationTemplateRepository: NotificationTemplateRepository,
     protected createExecutionDetails: CreateExecutionDetails
   ) {}
 
   async execute(command: TriggerEventCommand) {
-    await this.validateTransactionIdProperty(command.transactionId, command.organizationId, command.environmentId);
+    const { actor, environmentId, identifier, organizationId, to, userId } = command;
 
-    const template = await this.notificationTemplateRepository.findByTriggerIdentifier(
-      command.environmentId,
-      command.identifier
-    );
+    await this.validateTransactionIdProperty(command.transactionId, organizationId, environmentId);
 
     Sentry.addBreadcrumb({
       message: 'Sending trigger',
       data: {
-        triggerIdentifier: command.identifier,
+        triggerIdentifier: identifier,
       },
     });
 
+    const template = await this.notificationTemplateRepository.findByTriggerIdentifier(
+      command.environmentId,
+      command.identifier
+    );
+
+    /*
+     * Makes no sense to execute anything if template doesn't exist
+     * TODO: Send a 404?
+     */
+    if (!template) {
+      const message = 'Notification template could not be found';
+      Logger.error(message, LOG_CONTEXT);
+      throw new ApiException(message);
+    }
```

Comment:
Saving execution time and DB calls if we reach this scenario.

Reply:
The use-case `TriggerEvent` is used by the `TriggerHandlerQueueService` and throwing that error won't work for the case when the trigger queue events are processed, because there is no REST context anymore... so I feel like it should throw some generic runtime error and we should catch and wrap it in other places... wdyt?

Reply:
Not totally understand the point you explain. `ApiException` is a generic runtime error. We already throw something similar in the method `validateTransactionIdProperty` that validates the entry data. And the worker at the end it has its own context, though mostly of the information provided will come from the database. I am happy to discuss this privately if it is easier because I couldn't understand what you mean, sorry. 😓

Reply:
`APIException` extends `BadRequestException` which is `HttpException`, but we are not in that HTTP context anymore, so I think that this is not an appropriate exception here.

Reply:
🤦🏻 We have a bigger problem then. We would need to change the definition of ApiException and the use in the monorepo. The are more places where it is used under the queue context. Adding to technical debt.